lang/py/src/avro/io.py

```diff
@@ -580,16 +696,36 @@ def read_data(self, writers_schema, readers_schema, decoder):
     elif writers_schema.type == 'string':
       return decoder.read_utf8()
     elif writers_schema.type == 'int':
-      return decoder.read_int()
+      if hasattr(writers_schema, 'logical_type'):
+        if writers_schema.logical_type == constants.DATE:
+          return decoder.read_date_from_int()
+        elif writers_schema.logical_type == constants.TIME_MILLIS:
+          return decoder.read_time_millis_from_int()
```

Comment:
I think the date before the epoch datetime should be acceptable since it's still a valid the date. However, the time should be non-negative according to the spec: `The time-millis logical type represents a time of day, with no reference to a particular calendar, time zone or date`. The negative time value seems not making much sense.

The validation should probably occur in the validation function instead of write function to each responsibility separate and clear.